### PR TITLE
Ability to send default Headers and Query #311

### PR DESCRIPTION
**Pull request by IlyasYOY**
Tried to implement it https://github.com/Playtika/feign-reactive/issues/311 here.

There are some bad things I've done here to support query. May anybody give me a hint how to improve this? Maybe I have to rewrite `ReactiveHttpRequest` to support query without any problems.

Thank you for your work! Use this project almost every day!

### CODE REVIEW

**Comment by IlyasYOY** [comment]
It works, but I guess I gotta change the way I add query 😞

File: feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java [hunk 1]
```diff
@@ -26,64 +28,82 @@
 
 /**
  * An immutable reactive request to an http server.
- * 
+ *
  * @author Sergii Karpenko
  */
 public final class ReactiveHttpRequest {
```
Comment (kptfh) [inline feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java:35]: please avoid formatting of code. I'd like to se actual changes

File: feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java [hunk 2]
```diff
@@ -26,64 +28,82 @@
...
+    }
+
+    /* Method to invoke on the server. */
+    public String method() {
+        return methodMetadata.template().method();
+    }
+
+    public ReactiveHttpRequest withQuery(String key, String value) {
+        String query = uri.getQuery();
+        String keyValuePair = String.format(KEY_VALUE_PAIR_TEMPLATE, key, value);
```
Comment (kptfh) [inline feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java:71]: String.format is extremely ineffective method. Try to avoid it

**Review by kptfh — COMMENTED** [review]
In general nice PR

File: feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java [hunk 3]
```diff
@@ -26,64 +28,82 @@
...
+    public ReactiveHttpRequest(ReactiveHttpRequest request, Publisher<Object> body) {
+        this(request.methodMetadata, request.target, request.uri, request.headers, body);
+    }
+
+    /* Method to invoke on the server. */
+    public String method() {
+        return methodMetadata.template().method();
+    }
+
+    public ReactiveHttpRequest withQuery(String key, String value) {
```
Comment (kptfh) [inline feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java:69]: move out this method to interceptor

File: feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java (hunk 2 above)
Comment (IlyasYOY) [reply feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java]: replaced with concatenation

File: feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java (hunk 3 above)
Comment (IlyasYOY) [reply feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java]: Moved this out

**Comment by IlyasYOY** [comment]
Also added additional tests on multiple fields

File: feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java (hunk 1 above)
Comment (IlyasYOY) [reply feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java:36]: Added back line break

Should I change indentation back?

Comment (kptfh) [reply feign-reactor-core/src/main/java/reactivefeign/client/ReactiveHttpRequest.java:36]: let's revert this class at all as it actually has no changes

File: feign-reactor-spring-configuration/src/main/java/reactivefeign/spring/config/ReactiveFeignBasicConfigurator.java [hunk 4]
```diff
@@ -160,4 +189,24 @@ static ReactiveRetryPolicy configureRetryPolicyFromProperties(
 		}
 		return retryPolicy;
 	}
+
+	private static ReactiveHttpRequest reactiveHttpRequestWithQuery(ReactiveHttpRequest reactiveHttpRequest, String key, String value) {
```
Comment (kptfh) [inline feign-reactor-spring-configuration/src/main/java/reactivefeign/spring/config/ReactiveFeignBasicConfigurator.java:193]: I this it's better to move out all code related to interceptors to ReactiveHttpRequestInterceptors

Create 
```
addQueries(){

for (Map.Entry<String, List<String>> entry : config.getDefaultQueryParameters().entrySet()) {
				// Every Map entry is gonna belong to it's own interceptor
				resultBuilder = resultBuilder.addRequestInterceptor(reactiveHttpRequest -> {
					for (String value : entry.getValue()) {
						reactiveHttpRequest = reactiveHttpRequestWithQuery(reactiveHttpRequest, entry.getKey(), value);
					}
					return Mono.just(reactiveHttpRequest);
				});
			}

}
```
and 
```
addHeaders(){
for (Map.Entry<String, List<String>> entry : config.getDefaultRequestHeaders().entrySet()) {
				// Every Map entry is gonna belong to it's own interceptor
				resultBuilder = resultBuilder.addRequestInterceptor(reactiveHttpRequest -> {
					reactiveHttpRequest.headers().put(entry.getKey(), entry.getValue());
					return Mono.just(reactiveHttpRequest);
				});
			}
}
```

Comment (IlyasYOY) [reply feign-reactor-spring-configuration/src/main/java/reactivefeign/spring/config/ReactiveFeignBasicConfigurator.java]: Extracted to `ReactiveHttpRequestInterceptors`. Now it looks really good, thanks!
More over, now I know about this useful class, can use it somewhere else in my project!

**Comment by kptfh** [comment]
Great work

**Comment by kptfh** [comment]
just released with this extension 2.0.24

**Pull request merged**